presto-orc/src/main/java/com/facebook/presto/orc/reader/AbstractDecimalSelectiveStreamReader.java

```diff
@@ -0,0 +1,350 @@
+            dataStream.skip(dataToSkip);
+            scaleStream.skip(dataToSkip);
+        }
+        else {
+            dataStream.skip(items);
+            scaleStream.skip(items);
+        }
+    }
+
+    @Override
+    public int[] getReadPositions()
+    {
+        return outputPositions;
+    }
+
+    @Override
+    public void throwAnyError(int[] positions, int positionCount)
+    {
+    }
+
+    protected BlockLease newLease(Block block)
+    {
+        valuesInUse = true;
+        return ClosingBlockLease.newLease(block, () -> valuesInUse = false);
+    }
+
+    @Override
+    public Block getBlock(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
+        checkState(positionCount <= outputPositionCount, "Not enough values");
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (allNulls) {
+            return new RunLengthEncodedBlock(nullBlock, outputPositionCount);
+        }
+
+        boolean includeNulls = nullsAllowed && presentStream != null;
+
+        if (positionCount == outputPositionCount) {
+            Block block = makeBlock(positionCount, nullsAllowed, nulls, values);
+            nulls = null;
+            values = null;
+            return block;
+        }
+
+        long[] valuesCopy = createValuesCopy(positionCount);
+        boolean[] nullsCopy = null;
+
+        if (includeNulls) {
+            nullsCopy = new boolean[positionCount];
+        }
+
+        copyValues(positions, positionCount, valuesCopy, nullsCopy);
+
+        return makeBlock(positionCount, includeNulls, nullsCopy, valuesCopy);
+    }
+
+    @Override
+    public BlockLease getBlockView(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
+        checkState(positionCount <= outputPositionCount, "Not enough values");
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (allNulls) {
+            return newLease(new RunLengthEncodedBlock(nullBlock, outputPositionCount));
+        }
+
+        boolean includeNulls = nullsAllowed && presentStream != null;
+        if (positionCount != outputPositionCount) {
+            compactValues(positions, positionCount, includeNulls);
+        }
+
+        return newLease(makeBlock(positionCount, includeNulls, nulls, values));
+    }
+
+    abstract void copyValues(int[] positions, int positionsCount, long[] valuesCopy, boolean[] nullsCopy);
+
+    abstract Block makeBlock(int positionCount, boolean includeNulls, boolean[] nulls, long[] values);
+
+    abstract long[] createValuesCopy(int positionCount);
+
+    protected void compactValues(int[] positions, int positionCount, boolean compactNulls)
```

Comment:
I think this method won't work for long decimals because it needs to know how many entries per position there are in `values` array.

Reply:
will follow up on this

Reply:
agree, added code to handle 2 positions for Long and modified a test that will catch this.